src/optimization/abstract_model_store.jl

```diff
@@ -0,0 +1,91 @@
+abstract type AbstractModelStore end
+
+# Required methods for subtypes:
+# = initialize_storage!
+# - write_result!
+# - read_results
+# - write_optimizer_stats!
+# - read_optimizer_stats
+#
+# Each subtype must have a field for each instance of STORE_CONTAINERS.
+
+function Base.empty!(store::AbstractModelStore)
+    stype = typeof(store)
+    for (name, type) in zip(fieldnames(stype), fieldtypes(stype))
+        val = get_data_field(store, name)
+        try
+            empty!(val)
+        catch
+            @error "Base.empty! must be customized for type $stype or skipped"
+            rethrow()
+        end
+    end
+end
+
+get_data_field(store::AbstractModelStore, type) = getfield(store, type)
```

Comment:
Why do we have this instead of using `getproperty` or `getfield`?

Reply:
I don't remember.

Reply:
fixed